doc/design/high-level.md

```diff
@@ -0,0 +1,221 @@
+
+The `xcodeproj` rule has to make some assumptions about the data it gets, as the
+providers from other rules don't have the fidelity needed to perfectly recreate
+a similar Xcode target. rules_xcodeproj will expose providers and associated
+helper functions, to allow rules, including your own custom ones, to control how
+the `xcodeproj` generates targets. The goal being that a default,
+non-customized, project is as natural as possible. Rules that return
+rules_xcodeproj providers can choose to expose customization points, similar to
+the additional rules mentioned above, but that's not their primary purpose.
+
+## Attempts to match the outputs of `bazel build`
+
+By virtue of the [first goal](#only-the-xcodeproj-rule-is-necessary), most
+Xcode outputs will match those produced by `bazel build`. There are other areas
+though, e.g. dealing with hermeticity, debugging, and indexing, where additional
+care will need to be taken.
+
+## Native feeling projects
+
+Ideally, you shouldn't be able to tell that a project was generated with
+rules_xcodeproj. It should look and feel like any other Xcode project. This
+should also be the case regardless of which [build mode](#multiple-build-modes)
+is used.
+
+When that isn't the case, and there needs to be a deviation (mainly to [satisfy
+other constraints](#attempts-to-match-the-outputs-of-bazel-build)), then the
+generated project should deviate as little as possible. A [different build
+mode](#build-with-bazel-via-proxy) might be required to make the project feel
+more native.
+
+## Defined in `BUILD` files
+
+The building blocks of a rules_xcodeproj project generation is the `xcodeproj`
+rule and associated [customization rules](#additional-rules). Targets using
+these rules are defined in `BUILD` files, and generating a project happens by
+executing  `bazel run` on an `xcodeproj` target. There is no need to create
+additional configuration files, or to run additional commands.
+
+Some setups might require something more dynamic, in particular when using the
+focused project customization. For these cases the recommended approach, which
+we might supply some optional tools for, is to dynamically generate `.bzl` files
+with macros that create the required targets and use those in your `BUILD`
+files.
+
+## Multiple build modes
+
+The `xcodeproj` rule will allow specifying a build mode that should be used by
+the generated project. This will allow the project to build with Xcode instead
+of Bazel, if that is desired.
+
+Here are a few reasons one may want to build with
+Xcode instead of Bazel:
+
+- If a new, possibly beta, version of Xcode is released with a feature that
+  Build with Bazel doesn't support yet, because one of Bazel, rules_apple,
+  rules_swift, or rules_xcodeproj doesn't support it
+- To work around a bug in Bazel, rules_apple, rules_swift, or rules_xcodeproj
+- To compare the Bazel and Xcode build systems or build commands
+- As a step when migrating to Bazel
```

Comment:
Do you anticipate getting SwiftUI previews to work using BWB? If not, is that worth adding here?

Reply:
SwiftUI previews will work, but they will require that the build graph support them: targets will need to be frameworks or bundles in some way. While rules_xcodeproj won't modify the build graph in any way to do that for teams, we can mention what needs to be done, probably in a `--config` so it looks right for the project generation.

I mention something like this in some of the detailed sections, so I think it's covered?